### ISSUE TITLE
Notification rule is missing a root condition - Improve error message

### ISSUE THREAD

**Issue by amjadniazi48** [rainlab/notify-plugin#10]
![rootcondition](https://user-images.githubusercontent.com/2420029/43463384-e3c3b014-94f1-11e8-9910-f5f213227168.png)
getting this error message how to get rid of this, i have the root condition as shown above but it shows
Notification rule is missing a root condition!
i have just removed the userplus plugin and it is working fine now.

**Comment by saulo-silva** [comment]
same problem

**Comment by LukeTowers** [comment]
@saulo-silva are you using the UserPlus plugin as well?

**Comment by vdomah** [comment]
Also has got this. Seems it's caused by uncompleted deferred binding of rule_conditions. First I was getting only rule_conditions records without rule_host_id. I'm not sure, but seems I've fixed this by resaving the notification rule in backend
http://prntscr.com/lqie4u

**Comment by LukeTowers** [comment]
If someone can narrow down a cause and submit a PR to fix it that would be helpful.

**Comment by BahodurSaidov** [comment]
i had same issue with automatic user activation by sending mail...
Something wrong with UserPlus plugin, it seems that some notification rule is created by the plugin without any rule_conditions...
Temporary solution would be to disable UserPlus plugin and disable Rain.notify plugin, but enable back notify plugin , all from backend...
This worked for me... wasted couple of hours searching though

**Comment by w20k** [comment]
@BahodurSaidov, could you provide a simple example of how to reproduce it?

**Comment by BahodurSaidov** [comment]
@w20k, i think you should be able to reproduce with build 446 install User & UserPlus plugin. Then enable registration, from User settings select activated by user (send link to mail).
Also, during this issue, i was deleting test user several times from backend and re-registering so to get test activation mail...
It should be mentioned that when the problem exists, the mail with activation link is NOT sent after registration form, and an exception is thrown at frontend with flash messaging saying something with "notification rule...".
This much I remember... Hope it helps...
Please let know if you could reproduce the issue by following these steps...

**Comment by w20k** [comment]
Side note: by default, if you use the built-in notification rule - email template is wrong.
@BahodurSaidov, mail template naming should fix your issue.

**Comment by web-studio-51** [comment]
Hi! Im from Russia, but you can translate my text:

Проблема решается очень просто, заходим в файл plugins/rainlab/notify/models/NotificationRule.php

и комментируем строки    71-77

   // if (!$rootCondition) {
        //     throw new ApplicationException('Notification rule is missing a root condition!');
        // }

        // if (!$rootCondition->getConditionObject()->isTrue($params)) {
        //     return false;
        // }

Отключив эту проверку, ошибка высвечиваться не будет, но плагин работает корректно

**Comment by BahodurSaidov** [comment]
@web-studio-51, интерестно зачем вообще добавили ApplicationException туда ?!

**Comment by w20k** [comment]
@web-studio-51, ты можешь описать, как ошибку зарепродюсить?

**Comment by w20k** [comment]
Wait, @LukeTowers, could you clarify one thing. Should notify-plugin rules work without defining a single condition? 
Finally understood what's the issue is present here 😃

**Comment by LukeTowers** [comment]
No idea @w20k, I haven't actually used this plugin

**Comment by w20k** [comment]
Darn, @LukeTowers, same for me. Used it few times, but with rules.

**Comment by w20k** [comment]
@amjadniazi48, @saulo-silva, @vdomah, @web-studio-51, @BahodurSaidov, for your information.

Will be closing this ticket. As I've understood from the code base, the plugin doesn't support cases where only a root condition is enabled. 
You need to specify one or more conditions.

And if it's not the case, please open a new ticket with description and steps to reproduce.

As for hacks: @web-studio-51
If you want to have this supported in the core of this plugin - Feel free to make a PR. 
Or an error showing that single conditions should be added.

**Comment by LukeTowers** [comment]
@w20k I think this needs to be corrected with an improvement to the error message. In this case the user has a root condition, but no actual subconditions to evaluate, thus the message is incorrect and confusing.

**Comment by w20k** [comment]
@LukeTowers, I was thinking to add additional validation based on the number of rules or if the repeater is empty. 
I agree! The message is definitely confusing. I've added this task to my ToDo List.

**Comment by LukeTowers** [comment]
Will re-open until the message gets improved.

**Comment by Balex70** [comment]
> i had same issue with automatic user activation by sending mail...
> Something wrong with UserPlus plugin, it seems that some notification rule is created by the plugin without any rule_conditions...
> Temporary solution would be to disable UserPlus plugin and disable Rain.notify plugin, but enable back notify plugin , all from backend...
> This worked for me... wasted couple of hours searching though

Faced problem with error message "Notification rule is missing a root condition", tried solution explained above. But turned out, in this case, I can't create notification rule at all. After enabling UserPlus again and hit "Notification rules" in settings area I got fresh default rule with wrong template naming "welcome". And I can't create a new rule because of that.
And every time I hit "Notification rules" in settings I got fresh default rule over and over again. Something similar was discussed here: https://github.com/rainlab/notify-plugin/issues/3

Tried to reinstall all related plugins: UserPlus, Location and Notify. Not helped.

Environment:
Ubuntu 17.10
php 7.1
Build: 446

**Comment by gusdgg** [comment]
I also get this error message "Notification rule is missing a root condition". In my case, I found that I had only one rule in the "Notification rules" settings with root condition but two records in the "rainlab_notify_rule_conditions" table with the same class name. So, when the Notifier class fires the event it founds two records, one with root condition but another one without root condition and triggers the error. I think that one of the records was still there because a failed delete action or something. When I deleted this record the issue was fixed.
